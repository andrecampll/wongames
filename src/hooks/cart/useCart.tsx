import { useContext, createContext, ReactNode } from 'react';

export type CartContextDTO = Record<string, never>;

export const CartContextDefaultValues = {};

const CartContext = createContext<CartContextDTO>(CartContextDefaultValues);

type Props = {
  children: ReactNode;
};

const CartProvider = ({ children }: Props) => {
  return <CartContext.Provider value={{}}>{children}</CartContext.Provider>;
};

const useCart = () => useContext(CartContext);

export { CartProvider, useCart };
