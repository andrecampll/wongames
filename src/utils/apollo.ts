import { useMemo } from 'react';
import { setContext } from '@apollo/client/link/context';
import {
  ApolloClient,
  HttpLink,
  NormalizedCache,
  NormalizedCacheObject,
} from '@apollo/client';
import { Session } from 'next-auth';
import apolloCache from './apolloCache';

let apolloClient: ApolloClient<NormalizedCacheObject | null>;

function createApolloClient(session?: Session | null) {
  const httpLink = new HttpLink({
    uri: `${process.env.NEXT_PUBLIC_API_URL}/graphql`,
  });

  const authLink = setContext((_, { headers, session: clientSession }) => {
    const jwt = session?.jwt || clientSession?.jwt || '';
    const authorization = jwt ? `Bearer ${jwt}` : '';

    return { headers: { ...headers, authorization } };
  });

  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    link: authLink.concat(httpLink),
    cache: apolloCache,
  });
}

export function initializeApollo(
  initialState = null,
  session?: Session | null,
) {
  const apolloClientGlobal = apolloClient ?? createApolloClient(session);

  if (initialState) {
    apolloClientGlobal.cache.restore(
      initialState as NormalizedCacheObject & NormalizedCache,
    );
  }

  if (typeof window === 'undefined') return apolloClientGlobal;

  apolloClient = apolloClient ?? apolloClientGlobal;

  return apolloClient;
}

export function useApollo(initialState = null, session?: Session) {
  const store = useMemo(() => initializeApollo(initialState, session), [
    initialState,
    session,
  ]);

  return store;
}
