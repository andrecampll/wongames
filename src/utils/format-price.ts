export function formatPrice(price: number | bigint): string {
  return new Intl.NumberFormat('en', {
    style: 'currency',
    currency: 'USD',
  }).format(price);
}
