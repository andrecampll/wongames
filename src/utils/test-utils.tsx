import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';

import {
  WishlistContext,
  WishlistContextDefaultValues,
  WishlistContextDTO,
} from '../hooks/wishlist/useWishlist';

import theme from '../styles/theme';
import {
  CartContext,
  CartContextDTO,
  CartContextDefaultValues,
} from '../hooks/cart/useCart';

type CustomRenderProps = {
  cartProviderProps?: CartContextDTO;
  wishlistProviderProps?: WishlistContextDTO;
} & Omit<RenderOptions, 'queries'>;

const customRender = (
  ui: ReactElement,
  {
    cartProviderProps = CartContextDefaultValues,
    wishlistProviderProps = WishlistContextDefaultValues,
    ...renderOptions
  }: CustomRenderProps = {},
) =>
  render(
    <ThemeProvider theme={theme}>
      <CartContext.Provider value={cartProviderProps}>
        <WishlistContext.Provider value={wishlistProviderProps}>
          {ui}
        </WishlistContext.Provider>
      </CartContext.Provider>
    </ThemeProvider>,
    renderOptions,
  );

export * from '@testing-library/react';
export { customRender as render };
