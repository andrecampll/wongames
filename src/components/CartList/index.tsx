import Link from 'next/link';
import { useCart } from '../../hooks/cart/useCart';

import Button from '../Button';
import Empty from '../Empty';
import GameItem from '../GameItem';
import Loader from '../Loader';

import { Wrapper, Footer, Total, Loading, GamesList } from './styles';

export type CartListProps = {
  hasButton?: boolean;
};

const CartList = ({ hasButton = false }: CartListProps) => {
  const { items, total, loading } = useCart();

  if (loading) {
    return (
      <Loading>
        <Loader />
      </Loading>
    );
  }

  return (
    <Wrapper isEmpty={!items.length}>
      {items.length ? (
        <>
          <GamesList>
            {items.map(item => (
              <GameItem key={item.title} {...item} />
            ))}
          </GamesList>

          <Footer>
            {!hasButton && <span>Total:</span>}
            <Total>{total}</Total>

            {hasButton && (
              <Link href="/cart">
                <Button as="a">Buy it now</Button>
              </Link>
            )}
          </Footer>
        </>
      ) : (
        <Empty
          title="Your cart is empty"
          description="Go back to the store and explore great games and offers."
          hasLink
        />
      )}
    </Wrapper>
  );
};

export default CartList;
