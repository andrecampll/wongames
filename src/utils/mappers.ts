import { QueryGames_games } from '../graphql/generated/QueryGames';
import {
  QueryHome_banners,
  QueryHome_sections_freeGames_highlight,
} from '../graphql/generated/QueryHome';
import { QueryWishlist_wishlists_games } from '../graphql/generated/QueryWishlist';

import { formatPrice } from './format-price';

export const bannerMapper = (banners: QueryHome_banners[]) => {
  return banners.map(({ image, title, subtitle, button, ribbon }) => ({
    image: `http://localhost:1337${image?.url}`,
    title,
    subtitle,
    buttonLabel: button?.label,
    buttonLink: button?.link,
    ...(ribbon && {
      ribbon: ribbon.text,
      ribbonColor: ribbon.color,
      ribbonSize: ribbon.size,
    }),
  }));
};

export const gamesMapper = (games: QueryGames_games[] | null | undefined) => {
  return games
    ? games.map(({ id, name, slug, developers, cover, price }) => ({
        id,
        title: name,
        slug,
        developer: developers[0].name,
        image: cover?.url
          ? `http://localhost:1337${cover?.url}`
          : 'https://i.stack.imgur.com/y9DpT.jpg',
        price,
      }))
    : [];
};

export const highlightMapper = (
  highlight: QueryHome_sections_freeGames_highlight | null | undefined,
) => {
  return highlight
    ? {
        title: highlight.title,
        subtitle: highlight.subtitle,
        backgroundImage: `http://localhost:1337${highlight.background.url}`,
        floatImage: `http://localhost:1337${highlight.floatImage.url}`,
        buttonLabel: highlight.buttonLabel,
        buttonLink: highlight.buttonLink,
        alignment: highlight.alignment,
      }
    : {};
};

export const cartMapper = (
  games:
    | QueryGames_games[]
    | QueryWishlist_wishlists_games[]
    | null
    | undefined,
) => {
  return games
    ? games.map(game => ({
        id: game.id,
        img: `http://localhost:1337${game.cover?.url}`,
        title: game.name,
        price: formatPrice(game.price),
      }))
    : [];
};
