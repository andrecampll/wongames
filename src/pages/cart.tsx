import { GetServerSidePropsContext } from 'next';
import protectedRoutes from '../utils/protectedUtils';
import { initializeApollo } from '../utils/apollo';
import { gamesMapper, highlightMapper } from '../utils/mappers';
import { QUERY_RECOMMENDED } from '../graphql/queries/recommended';
import { QueryRecommended } from '../graphql/generated/QueryRecommended';

import Cart, { CartPageProps } from '../templates/Cart';

export default function CartPage(props: CartPageProps) {
  return <Cart {...props} />;
}

export async function getServerSideProps(ctx: GetServerSidePropsContext) {
  const session = await protectedRoutes(ctx);
  const apolloClient = initializeApollo(null, session);

  const { data } = await apolloClient.query<QueryRecommended>({
    query: QUERY_RECOMMENDED,
  });

  return {
    props: {
      session,
      recommendedTitle: data.recommended?.section?.title,
      recommendedGames: gamesMapper(data.recommended.section.games),
      recommendedHighlight: highlightMapper(data.recommended.section.highlight),
    },
  };
}
