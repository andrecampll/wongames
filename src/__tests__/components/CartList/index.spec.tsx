import { screen } from '@testing-library/react';
import { render } from '../../../utils/test-utils';
import items from '../../../components/CartList/mock';

import CartList from '../../../components/CartList';
import { CartContextDefaultValues } from '../../../hooks/cart/useCart';

describe('CartList', () => {
  it('should be able to render', () => {
    const cartProviderProps = {
      ...CartContextDefaultValues,
      items,
      total: 'R$ 330,00',
    };

    const { container } = render(<CartList />, {
      cartProviderProps,
    });

    expect(screen.getAllByRole('heading')).toHaveLength(2);
    expect(screen.getByText('R$ 330,00')).toHaveStyle({ color: '#F231A5' });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render loading', () => {
    const cartProviderProps = {
      ...CartContextDefaultValues,
      loading: true,
    };

    render(<CartList hasButton />, { cartProviderProps });

    expect(screen.getByTitle(/loading/i)).toBeInTheDocument();
  });

  it('should be able the button', () => {
    const cartProviderProps = {
      ...CartContextDefaultValues,
      items,
    };

    render(<CartList hasButton />, {
      cartProviderProps,
    });

    expect(screen.getByText(/buy it now/i)).toBeInTheDocument();
  });

  it('should render empty if there are no games', () => {
    render(<CartList />);

    expect(screen.getByText(/your cart is empty/i)).toBeInTheDocument();
    expect(screen.queryByText(/total/i)).not.toBeInTheDocument();
  });
});
