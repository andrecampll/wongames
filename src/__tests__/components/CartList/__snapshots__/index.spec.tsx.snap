// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartList should be able to render 1`] = `
<main
  class="sc-kstrdz gXjLku"
>
  <div
    class="sc-hBEYos ieNNDx"
  >
    <div
      class="sc-jSgupP hnZGpM"
    >
      <div
        class="sc-gKsewC jSeLHy"
      >
        <div
          class="sc-iBPRYJ bZbbiG"
        >
          <img
            alt="Red Dead Redemption 2"
            src="https://source.unsplash.com/user/willianjusten/151x70"
          />
        </div>
        <div
          class="sc-fubCfw bpBGEF"
        >
          <h3
            class="sc-kEjbxe gNpxNo"
          >
            Red Dead Redemption 2
          </h3>
          <div
            class="sc-pFZIQ lppYwz"
          >
            <p
              class="sc-iqHYGH cjANlT"
            >
              R$ 215,00
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-jSgupP hnZGpM"
    >
      <div
        class="sc-gKsewC jSeLHy"
      >
        <div
          class="sc-iBPRYJ bZbbiG"
        >
          <img
            alt="Borderlands 3"
            src="https://source.unsplash.com/user/willianjusten/151x70"
          />
        </div>
        <div
          class="sc-fubCfw bpBGEF"
        >
          <h3
            class="sc-kEjbxe gNpxNo"
          >
            Borderlands 3
          </h3>
          <div
            class="sc-pFZIQ lppYwz"
          >
            <p
              class="sc-iqHYGH cjANlT"
            >
              R$ 215,00
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="sc-fFubgz hfMkoR"
  >
    <span>
      Total:
    </span>
    <span
      class="sc-bkzZxe dJFdXZ"
    >
      R$ 330,00
    </span>
  </div>
</main>
`;
